feat: add support for remaining args

MangelMaxime opened the pull request:
I don't know how to make this change testable because it happens at the creation of the `Pipeline` so I added a demo script to showcase what it does and as an easy way to check that all is good for you.

One idea for making it testable was to change:

https://github.com/slaveOftime/Fun.Build/blob/3053b5eb9fa259ae5d6a454e6b2aa375fbb2544f/Fun.Build/PipelineContextExtensions.fs#L13-L15

to 

```fs
PipelineContext with 
  
     static member Create(name: string, ?forceArgs : string array) =
```

but I don't think this is good idea to change an API just to make it testable. Because I don't think we ever want to override the `Environment.GetCommandLineArgs()` in a normal usage.

albertwoo commented on Fun.Build/StageContextExtensions.fs:
```diff
@@ -499,6 +499,23 @@ module StageContextExtensions =
+        member ctx.GetRemainingsCmdArgs() =
+            match ctx.ParentContext with
+            | ValueSome(StageParent.Pipeline p) -> p.RemainingCmdArgs
+            | ValueSome(StageParent.Stage s) -> s.GetRemainingsCmdArgs()
```
Is the "s" after Remaining necessary? My English is not well to know that. But I see in the PipelineContext you are not using s for **RemainingCmdArgs**, so just wandering.

MangelMaxime replied:
Good catch.

I checked and Spectre.Console.Cli use `Remaining` for the name of property.

One less letter to type is always good 😊

albertwoo commented on Fun.Build/StageContextExtensions.fs:
```diff
@@ -499,6 +499,23 @@ module StageContextExtensions =
             | ValueSome(StageParent.Stage s) -> s.GetAllCmdArgs()
```
Can we add some comments on the **GetAllCmdArgs** to identify it will not return the remaing ones?

albertwoo commented:
@MangelMaxime for the unit test, I think you can write code like:

```fsharp
let mutable actualAllCmdArgs = []
let mutable actualRemainingArgs = []

pipeline "demo" {
    cmdArgs ["-p"; "demo"; "test1"; "v1"; "--"; "test2"; "v2"]
    stage "shows arguments" {
        run (fun ctx ->
            actualAllCmdArgs  <- ctx.GetAllCmdArgs()
            actualRemainingArgs  <- ctx.GetRemainingsCmdArgs()
        )
    }
    runImmediate
}

// Assert accordingly

```

MangelMaxime commented:
Regarding the unit test, originally I wrote something like you suggested but in the middle of writing this PR I changed how I implemented `RemainingArgs` and removed the tests because I think `cmdArgs` take effect after `PipelineContext.Create` has been called.

I will give it a try and check what happens.

MangelMaxime commented:
I checked and with my current implementation I can't test it the way you proposed.

However, looking at the comment in the code, I think I should also add the Args/Remaining split logic at this place too no?

https://github.com/slaveOftime/Fun.Build/blob/bedb12fedab68204d9b3985c9fe34dcea94272d9/Fun.Build/PipelineBuilder.fs#L174-L181

albertwoo commented:
Yes, I think we should do that.

MangelMaxime commented:
@albertwoo This is ready for review.

By using making `cmdArgs` support Remaining args too and using an utility type to transform a list of args into a list of args + remaining args, I am able to indirectly test both places where this transformation takes place.

MangelMaxime commented:
The failing test doesn't seem related to my changes but caused by a slow CI runner.

albertwoo commented:
The failing test is related to a paralle tests, existing before, but did not figure out why 😂.

I will merge your PR, thanks for your contribution.
